File: scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def sum(factors: Set[Int], limit: Int): Int = {
    factors.flatMap(e => {
      for {
        i <- Range(1, limit + 1)
        candi = e * i
        if (candi < limit)
      } yield candi
    }).sum
  }
}

File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.List
import qualified Data.Set as Set

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum . Set.unions $ map (\x -> Set.fromAscList $ takeWhile (< limit) [x * y | y <- [1..]]) effectiveFactors
    where
        effectiveFactors = filter (/= 0) factors
